### DISCUSSION TITLE
Expected Behavior if Source Repository becomes inaccessible

### DISCUSSION THREAD

**Discussion by pinkfloydx33** [fluxcd/flux2#3452] [Q&A]
What is the expected behavior of Flux if the source repository suddenly becomes unavailable (networking issues, public key expires/changes) with Kustomizations set to `prune=true`?

We are running Flux on AKS via the "Gitops" portal-pane and the Flux v2 AKS extension. We are connecting to a Git source repository in Azure Devops using SSH and the public key generated by the extension (from Flux). The other day, the public key on one of my Flux installs apparently automatically rotated/updated. Because of this the source repository became inaccessible and everything that was part of a Kustomization set to prune was removed from my cluster.

Is this expected to happen? I don't know why the public key changed, it was meant to expire in late 2023. It only happened to one cluster (three others all bootstrapped around the same time were fine). Of course this feels like something wrong with the Flux extension in Azure and I intend to open a ticket with them. But that aside, should everything have pruned itself? I'd have at least expected something of a grace period. Since the generated public keys seem to be only good for ~18months, how do you go about rotating them without pruning everything? 

Or does this sound like something **really** went wrong? Should an inaccessible source repository cause such issues? Or should it keep all Kustomizations until its able to actually connect and do a real reconciliation?

Thankfully this happened in my QA cluster over the Christmas weekend and everyone happened to be out of office this week....once I figured out that the key had changed (by mere luck) I was able to update the security in Azure DevOps, force a reconcile and everything was reinstalled.... It wasn't prod, but man it could've been really bad. For now I've gone and set all my Kustomizations on all clusters so that they won't prune. 😢

**Comment by kingdonb** [comment]
> Is this expected to happen?

No, it is definitely not expected to happen and thank you for reporting the bad experience. When a source becomes unavailable, the Kustomization should not prune based on unavailability of the source, that is definitely not expected behavior in any way.

There should be no effect on your deployed workloads when a key goes out of service, except that they stop syncing. A Flux Kustomization should only prune or garbage collect when the source is actually available and the content of the target path synced is empty *and exists*.

That's very strange behavior indeed.

> Of course this feels like something wrong with the Flux extension in Azure and I intend to open a ticket with them.

That is what I was going to suggest you do. I work with the Azure team indirectly and I am also a Flux maintainer, I'd be very interested to hear what they tell you. Welcome to contact me directly on the CNCF Slack.

I am not sure of any way how this can happen, but unless I have just read your posting more than once, you are not the first person I've heard say this happened to them recently. I don't have an explanation for it now, and I'm very sorry this happened. We'll get to the bottom. Right now a lot of Flux developers are on holiday, but we'll definitely be having a look at this soon. Thanks again for reporting here.

Can you please attach the output of `flux check` here? I know you are on Azure's Flux distribution, but we should still be able to tell what version of the controllers were in use, although if the cluster has had its content nuked, I'm not sure if we can rely on that information anymore... (unless you can actually reproduce the issue, in which case we'll be **very** interested...)
